src/app/Library/CrudPanel/Traits/Input.php

```diff
@@ -0,0 +1,195 @@
+{
+    /**
+     * ---------------
+     * PRIVATE METHODS
+     * ---------------.
+     */
+
+    /**
+     * Returns the direct inputs parsed for model and relationship creation.
+     *
+     * @param  array  $inputs
+     * @param  null|array  $relationDetails
+     * @param  bool|string  $relationMethod
+     * @return array
+     */
+    private function splitInputIntoDirectAndRelations($inputs, $relationDetails = null, $relationMethod = false)
+    {
+        $crudFields = $relationDetails['crudFields'] ?? [];
+        $model = $relationDetails['model'] ?? false;
+
+        $directInputs = $this->getDirectInputsFromInput($inputs, $model, $crudFields, $relationMethod);
+        $relationInputs = $this->getRelationDetailsFromInput($inputs, $crudFields, $relationMethod);
+
+        return [$directInputs, $relationInputs];
+    }
+
+    /**
+     * Returns the attributes with relationships stripped out from the input.
+     * BelongsTo relations are ensured to have the correct foreign key set.
+     * ALL other relations are stripped from the input.
+     *
+     * @param  array  $input  - the input array
+     * @param  mixed  $model  - the model of what we want to get the attributtes for
+     * @param  array  $fields  - the fields used in this relation
+     * @param  mixed  $relationMethod  - the relation method
+     * @return array
+     */
+    private function getDirectInputsFromInput($input, $model = false, $fields = [], $relationMethod = false)
+    {
+        $model = $model ? (is_string($model) ? app($model) : $model) : $this->model;
+
+        $input = $this->decodeJsonCastedAttributes($input, $model);
+        $input = $this->compactFakeFields($input);
+        $input = $this->excludeRelationFieldsExceptBelongsTo($input, $fields, $relationMethod);
+        $input = $this->changeBelongsToNamesFromRelationshipToForeignKey($input, $fields);
+
+        return $input;
+    }
+
+    /**
+     * Get a relation data array from the form data. For each relation defined in the fields
+     * through the entity attribute, and set some relation details.
+     *
+     * We traverse this relation array later to create the relations, for example:
+     * - Current model HasOne Address
+     * - Address (line_1, country_id) BelongsTo Country through country_id in Address Model
+     *
+     * So when editing current model crud user have two fields
+     * - address.line_1
+     * - address.country
+     * (we infer country_id from relation)
+     *
+     * Those will be nested accordingly in this relation array, so address relation
+     * will have a nested relation with country.
+     *
+     * @param  array  $input  The form input.
+     * @param  array  $crudFields  - present when getting the relation details for other relations.
+     * @param  mixed  $relationMethod
+     * @return array The formatted relation details.
+     */
+    private function getRelationDetailsFromInput($input, $crudFields = [], $relationMethod = false)
```

Comment:
I've been trying to understand the structure of `$relationDetails` (finally!) and I think I'm close to understanding it. The name certainly hasn't helped, I think a more appropriate name would be `$relationTree` or `$relationHierarchy` - right?

AFAIK, the difference between this and fields is that this is nested - each relationship inside its parent.

Let's have a quick debugging session, to understand its structure. I feel we can achieve more in a call. Maybe we change its structure now, maybe we leave it like it is... but right now it seems a bit disorganized, I think one-two changes will greatly improve its readability.

Thanks!